### PR TITLE
Update APISPEC.md

### PR DESCRIPTION
**Pull request by Merkush1**
Instead of updating the documentation, I decided to rewrite it in a little more detail in a standardized format. It turned out a little voluminous, I know, I tried to cram as much as possible under expandable content. Better to check out it in visual format. Happy to have some comments from you, i can update to more shorter format or rechange spec. 

Also i had some problems with few endpoints:

1. POST /activity/delete - tried to test it many times but i can't figure it out. Tested it several times, but I still do not understand what is in the body of the request, looking at the code, so this endpoint is poorly described, as soon as we figure it out, I'll update it before the pr

2. Some strange behavior with several endpoints, I'll leave some comments in controllers

### CODE REVIEW

**Review by DrVilepis — APPROVED** [review]
Holy poop this is perfect, we will merge this as soon as possible.

Thank you for this amazing contribution!

**Comment by DrVilepis** [comment]
/activity/delete is in fact not a POST endpoint, its a DELETE, and it currently just takes a single integer, which is the id of a activity. Its kind of a black sheep of the endpoints as it does not take json as input, even though it should. We will fix this in the next release, as this is not the intended way for it to take input.

File: docs/APISPEC.md [hunk 1]
```diff
@@ -2,221 +2,1237 @@
...
+| --- | --- | --- | 
+| Authorized user is not part of found leaderboard or user is not an admin | 401 Unauthorized | { "error": "You are not authorized"} |
+| Demoting user is not the leaderboard member | 403 Forbidden | { "error": "You're not a member"} |
+</details>
+
+#### <a name="kick_lb"></a>  [9. POST /leaderboards/{name}/kick](#leaderboards)
+
+Kicks user from leaderboard if authorized user has admin rights
+
+>*This request is idempotent, it means that you can demote user that is already regular and have in response 200 OK*
```
Comment (DrVilepis) [inline docs/APISPEC.md:1188]: This does not seem like this should be here, should be an easy fix.

**Review by DrVilepis — CHANGES_REQUESTED** [review]

**Comment by DrVilepis** [comment]
Could you also possibly write a quick guide/explanation on what/how to document in CONTRIBUTING.md for easier documentation for us/other new contributors in the future.

**Comment by Merkush1** [comment]
>Could you also possibly write a quick guide/explanation on what/how to document in CONTRIBUTING.md for easier documentation for us/other new contributors in the future.

Yeah, sure! I'll check out about POST /activity/delete and other comments tomorrow ;)

**Comment by DrVilepis** [comment]
Thank you! I will be eagerly waiting.

**Comment by Merkush1** [comment]
>/activity/delete is in fact not a POST endpoint, its a DELETE, and it currently just takes a single integer, which is the id of a activity

Still getting the same problem. 

1) First request to get activity_id:

```
curl --location --request GET 'https://testaustime.fi/api/users/@me/activity/data' \
--header 'Authorization: Bearer <token>t' | python3 -m json.tool

[
    {
        "id": 39001,
        "start_time": "2022-08-07T16:06:58.154974",
        "duration": 19,
        "project_name": "string1",
        "language": "key1",
        "editor_name": "string",
        "hostname": "string"
    }
]
```

2) Second request to delete:

```
curl --location --request DELETE 'https://testaustime.fi/api/activity/delete' \
--header 'Authorization: Bearer <token>' \
--header 'Content-Type: text/plain' \
--data-raw '39001' -i

HTTP/1.1 404 Not Found
Server: nginx/1.14.2
Date: Sat, 13 Aug 2022 10:21:58 GMT
Content-Length: 0
Connection: keep-alive
ratelimit-limit: 8
vary: Origin, Access-Control-Request-Method, Access-Control-Request-Headers
ratelimit-reset: 10
ratelimit-remaining: 4
```

File: docs/APISPEC.md [hunk 2]
```diff
@@ -2,221 +2,1225 @@
...
-- Ratelimit: 10 req/m
+Testaustime API gives 5 different routes:
+- [/auth/](#auth)
+- [/users/](#users)
+- [/activity/](#activity)
+- [/friends/](#friends)
+- [/leaderboards/](#leaderboards)
 
-The desired interval at which to send heartbeats is immediately when editing a file, and after that at max every 30 seconds, and only when the user does something actively in the editor.
+Basic path: `https://testaustime.fi/api`
```
Comment (lajp) [inline docs/APISPEC.md:12]: The path should be https://api.testaustime.fi/ instead.
Same path should be used throughout the documentation.

File: docs/APISPEC.md [hunk 3]
```diff
@@ -2,221 +2,1225 @@
...
+  
+| Param |  Type | Required | Description |
+| --- | --- | --- | --- |
+| username | string | Yes | Usename has to be between 2 and 32 characters long |
+| password | string | Yes | Password has to be between 8 and 128 characters long |
+</details>
+
+**Sample request**
+```curl
+curl --request POST https://testaustime.fi/api/auth/register' \
```
Comment (lajp) [inline docs/APISPEC.md:54]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 4]
```diff
@@ -2,221 +2,1225 @@
...
+  
+| Param |  Type | Required | Description |
+| --- | --- | --- | --- |
+| username | string | Yes | Usename has to be between 2 and 32 characters long |
+| password | string | Yes | Password has to be between 8 and 128 characters long |
+</details>
+
+**Sample request**
+```curl
+curl --request POST 'https://testaustime.fi/api/auth/login' \
```
Comment (lajp) [inline docs/APISPEC.md:105]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 5]
```diff
@@ -2,221 +2,1225 @@
...
+  <summary>Body params:</summary>
+  
+| Param |  Type | Required | Description |
+| --- | --- | --- | --- |
+| new | string | Yes | New username. Usename has to be between 2 and 32 characters long |
+</details>
+
+**Sample request**
+```curl
+curl --request POST 'https://testaustime.fi/api/auth/changeusername' \
```
Comment (lajp) [inline docs/APISPEC.md:159]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 6]
```diff
@@ -2,221 +2,1225 @@
...
+| Param |  Type | Required | Description |
+| --- | --- | --- | --- |
+| old | string | Yes | Current password |
+| new | string | Yes | New password. Password has to be between 8 and 128 characters long |
+</details>
 
-Returns:
+**Sample request**
+```curl
+curl --request POST 'https://testaustime.fi/api/auth/changepassword' \
```
Comment (lajp) [inline docs/APISPEC.md:203]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 7]
```diff
@@ -2,221 +2,1225 @@
...
+  <summary>Header params:</summary>
+  
+| Name |  Value | 
+| --- | --- | 
+| Authorization | Bearer `<token>` |
+</details>
+
+**Sample request**
+```curl
+curl --request POST 'https://testaustime.fi/api/auth/regenerate' \
```
Comment (lajp) [inline docs/APISPEC.md:240]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 8]
```diff
@@ -2,221 +2,1225 @@
...
+  <summary>Header params:</summary>
+  
+| Name |  Value | 
+| --- | --- | 
+| Authorization | Bearer `<token>` |
+</details>
+
+**Sample request**
+```curl
+curl --location --request GET 'https://testaustime.fi/api/users/@me' \
```
Comment (lajp) [inline docs/APISPEC.md:286]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 9]
```diff
@@ -2,221 +2,1225 @@
...
+  
+| Name |  Value | 
+| --- | --- | 
+| Authorization | Bearer `<token>` |
+</details> 
 
-Required headers:
+**Sample request**
+```curl
+curl --location --request GET 'https://testaustime.fi/api/users/@me/leaderboards' \
```
Comment (lajp) [inline docs/APISPEC.md:324]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 10]
```diff
@@ -2,221 +2,1225 @@
...
+  
+| Path param |  Description | 
+| --- | --- | 
+| Username | Own or friend username. Also own username can be replaced on `@me`|
+</details> 
 
-Accepts:
+**Sample request**
+```curl
+curl --location --request GET 'https://testaustime.fi/api/users/@me/activity/data' \
```
Comment (lajp) [inline docs/APISPEC.md:369]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 11]
```diff
@@ -2,221 +2,1225 @@
...
+  
+| Param |  Type | Required | Description |
+| --- | --- | --- | --- |
+| username| string | Yes | Username |
+| password | string | Yes | User password |
+</details> 
+
+**Sample request**
+```curl
+curl --request DELETE 'https://testaustime.fi/api/users/@me/delete' \
```
Comment (lajp) [inline docs/APISPEC.md:425]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 12]
```diff
@@ -2,221 +2,1225 @@
...
+| editor_name | string | Name of IDA (Visual Studio Code, IntelliJ, Neovim, etc.) in which user is coding |
+| project_name | string| Name of the project in which user have a code session |
+</details> 
+
+**Sample first request**
+
+If the user doesn't have any active code session with this set of body params, then first request `POST /activity/update` creates new code session. Any other code session automatically stops/flushes after starting new one, so the user can't have >1 active code sessions in one time 
+
+```curl
+curl --request POST 'https://testaustime.fi/api/activity/update' \
```
Comment (lajp) [inline docs/APISPEC.md:481]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 13]
```diff
@@ -2,221 +2,1225 @@
...
+**Sample first response** 
+```HTTP
+200 OK
 ```
-<AUTHTOKEN>
+
+**Sample next request**
+
+```curl
+curl --request POST 'https://testaustime.fi/api/activity/update' \
```
Comment (lajp) [inline docs/APISPEC.md:500]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 14]
```diff
@@ -2,221 +2,1225 @@
...
+  <summary>Header params:</summary>
+  
+| Name |  Value | 
+| --- | --- | 
+| Authorization | Bearer `<token>` |
+</details> 
+
+**Sample request**
+```curl
+curl --request POST 'https://testaustime.fi/api/activity/flush' \
```
Comment (lajp) [inline docs/APISPEC.md:533]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 15]
```diff
@@ -2,221 +2,1225 @@
...
+  <summary>Body params:</summary>
+  
+| Param | Type | Description | 
+| --- | --- | --- | 
+| raw text | string | Activity id from response [`GET /users/{username}/activity/data`](#activity_data) |
+</details> 
+
+**Sample request** 
+```curl
+curl --request DELETE 'https://testaustime.fi/api/activity/delete' \
```
Comment (lajp) [inline docs/APISPEC.md:564]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 16]
```diff
@@ -2,221 +2,1225 @@
...
+  
+| Param | Type | Description | 
+| --- | --- | --- | 
+| raw text | string | Should contain friend code without any prefixes |
+</details> 
+
+**Sample request**
+
+```curl
+curl --request POST 'https://testaustime.fi/api/friends/add' \
```
Comment (lajp) [inline docs/APISPEC.md:610]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 17]
```diff
@@ -2,221 +2,1225 @@
...
+| Name |  Value | 
+| --- | --- | 
+| Authorization | Bearer `<token>` |
+</details> 
 
-Accepts:
+**Sample request**
+
+```curl
+curl --request GET 'https://testaustime.fi/api/friends/list' \
```
Comment (lajp) [inline docs/APISPEC.md:644]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 18]
```diff
@@ -2,221 +2,1225 @@
...
+  <summary>Header params:</summary>
+  
+| Name |  Value | 
+| --- | --- | 
+| Authorization | Bearer `<token>` |
+</details> 
+
+**Sample request**
+```curl
+curl --request POST 'https://testaustime.fi/api/friends/regenerate' \
```
Comment (lajp) [inline docs/APISPEC.md:688]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 19]
```diff
@@ -2,221 +2,1225 @@
...
+  <summary>Body:</summary>
+  
+| Param | Type | Description | 
+| --- | --- | --- | 
+| raw text | string | Should contain username without any prefixes |
+</details> 
+
+**Sample request** 
+```curl 
+--location --request DELETE 'https://testaustime.fi/api/activity/delete' \
```
Comment (lajp) [inline docs/APISPEC.md:729]: Why does the example not include the `curl` command itself.
The endpoint is also incorrect. Shouldn't it be `/friends/remove`?
Additionally change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 20]
```diff
@@ -2,221 +2,1225 @@
...
+  <summary>Body params:</summary>
+  
+| Param | Type | Description | 
+| --- | --- | --- | 
+| name| string | Name of creating leaderboard |
+</details> 
+
+**Sample request**
+```curl
+curl --request POST 'https://testaustime.fi/api/leaderboards/create' \
```
Comment (lajp) [inline docs/APISPEC.md:780]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 21]
```diff
@@ -2,221 +2,1225 @@
...
+| Param | Type | Description | 
+| --- | --- | --- | 
+| invite | string | Invite code for joining leaderboard |
+</details> 
+
+**Sample request**
 
-Returns:
+```curl
+curl --request POST 'https://testaustime.fi/api/leaderboards/join' \
```
Comment (lajp) [inline docs/APISPEC.md:834]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 22]
```diff
@@ -2,221 +2,1225 @@
...
+  
+| Path param | Description | 
+| --- | --- | 
+| {name} | Leaderboard name |
+</details> 
+
+**Sample request**
+
+```curl
+curl --request GET 'https://testaustime.fi/api/leaderboards/<name>' \
```
Comment (lajp) [inline docs/APISPEC.md:892]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 23]
```diff
@@ -2,221 +2,1225 @@
...
+
+| Path param | Description | 
+| --- | --- | 
+| {name} | Leaderboard name |
+</details> 
+
+**Sample request**
+
+```curl
+curl --request DELETE 'https://testaustime.fi/api/leaderboards/{name}' \
```
Comment (lajp) [inline docs/APISPEC.md:959]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 24]
```diff
@@ -2,221 +2,1225 @@
...
+
+| Path param | Description | 
+| --- | --- | 
+| {name} | Leaderboard name |
+</details> 
+
+**Sample request**
+
+```curl
+curl --request POST 'https://testaustime.fi/api/leaderboards/{name}/leave' \
```
Comment (lajp) [inline docs/APISPEC.md:1000]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 25]
```diff
@@ -2,221 +2,1225 @@
...
+
+| Path param | Description | 
+| --- | --- | 
+| {name} | Leaderboard name |
+</details> 
+
+**Sample request**
+
+```curl
+curl --request POST 'https://testaustime.fi/api/leaderboards/{name}/regenerate' \
```
Comment (lajp) [inline docs/APISPEC.md:1042]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 26]
```diff
@@ -2,221 +2,1225 @@
...
+
+| Param | Type | Description | 
+| --- | --- | --- | 
+| user | string | Username of a leaderboard member you want to promote |
+</details>
+
+**Sample request**
+
+```curl
+curl --request POST 'https://testaustime.fi/api/leaderboards/{name}/promote' \
```
Comment (lajp) [inline docs/APISPEC.md:1097]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 27]
```diff
@@ -2,221 +2,1225 @@
...
+| Param | Type | Description | 
+| --- | --- | --- | 
+| user | string | Username of a leaderboard admin you want to demote|
+</details>
 
+**Sample request**
 
-Required headers:
+```curl
+curl --request POST 'https://testaustime.fi/api/leaderboards/{name}/demote' \
```
Comment (lajp) [inline docs/APISPEC.md:1151]: change path to https://api.testaustime.fi

File: docs/APISPEC.md [hunk 28]
```diff
@@ -2,221 +2,1225 @@
...
+
+| Param | Type | Description | 
+| --- | --- | --- | 
+| user | string | Username of a leaderboard member you want to kick|
+</details>
+
+**Sample request**
+
+```curl
+curl --request POST 'https://testaustime.fi/api/leaderboards/{name}/kick' \
```
Comment (lajp) [inline docs/APISPEC.md:1206]: change path to https://api.testaustime.fi

**Review by lajp — CHANGES_REQUESTED** [review]
This is already looking very nice!
The main problem for me is that the apispec uses the path https://testaustime.fi/api when it should use https://api.testaustime.fi which is the official path.

There was also an issue with one of the sample requests which seemed to be wrong.
Fix these and we'll soon be able to merge this beast.

**Comment by Merkush1** [comment]
Done in new commit! Btw though still getting the same problem with activity/delete using right path `https://api.testaustime.fi` , still getting 404 as it was in https://github.com/Testaustime/testaustime-backend/pull/40#issuecomment-1214134465

**Comment by DrVilepis** [comment]
Yes. Turns out that was flawed routing from our side, that should be an easyish fix.

**Comment by Merkush1** [comment]
Hey @DrVilepis , is there anything else that i can fix in this PR, so you would merge this? ^^

**Comment by DrVilepis** [comment]
There isn't really anything left, we should be able to merge this the moment @lajp is available.

In fact the moment lajp sees this, he can feel free to merge this pr, this looks all good to me.

**Review by DrVilepis — APPROVED** [review]
Seems all good now

**Review by lajp — APPROVED** [review]
LGTM! Thanks once again for your contribution!

**Comment by Merkush1** [comment]
Thank you guys, At some point in time I'll make a swagger as a next step!

**Comment by bluelhf** [comment]
> Thank you guys, At some point in time I'll make a swagger as a next step!

A meme showcasing a photo of Ernest Khalimov, also known as GigaChad, with the caption "POV: You're Merkush1".

**Comment by DrVilepis** [comment]
I second this image

**Pull request merged**